File: tools/automation/src/mining-criteria-evidence.ts
import { createHash } from "node:crypto";
import { chmod, mkdir, readFile, rename, stat, unlink, writeFile } from "node:fs/promises";
import path from "node:path";

import type {
  AutomationCategory,
  MiningCriteriaEvidence,
  ProtocolCoordinatorEvidence,
} from "./contracts.generated.js";
import { internalCommandSpec } from "./contracts.generated.js";
import { optionValue, type ParsedInvocation } from "./invocation.js";
import type { ProcessPort } from "./process.js";
import { assertWithinWorkspace } from "./workspace.js";

export type MiningCriteriaEvidenceOptions = {
  readonly summary: string;
  readonly smoke: string;
  readonly soak: string;
  readonly coordinatorProjection: string;
  readonly projection: string;
};

export type MiningCriteriaValidators = {
  readonly coordinator: string;
  readonly evidence: string;
};

export type MiningCriteriaAdmittedDigests = {
  readonly summary: string;
  readonly smoke: string;
  readonly soak: string;
  readonly coordinator: string;
};

type FailureCategory = Extract<AutomationCategory, "evidence_invalid" | "process_failed">;
const sourceSpecifications = {
  summary: {
    path: "docs/parity/evidence/phase-21-live-mining-and-soak-evidence/summary.md",
    sha256: "b411ed3d8a1ce427231ec2818ed74fb590e6b29e4539a0e131bfdc7bc7acec0c",
  },
  smoke: {
    path: "docs/parity/evidence/phase-21-live-mining-and-soak-evidence/live-mining-smoke.md",
    sha256: "faec052c13b55cc7a53a1206c25c2094d93945d4b17d69c17c8a976e860655ff",
  },
  soak: {
    path: "docs/parity/evidence/phase-21-live-mining-and-soak-evidence/bounded-soak.md",
    sha256: "fc8904a9d9e2132789d70a9886c8aef05be96134e1ccd4d29bc793c9efa66003",
  },
  coordinator: {
    path: "docs/parity/evidence/str006-protocol-coordinator/protocol-coordinator-projection.json",
    sha256: "f008171f26b7a8ae6b08859e3cfef4f0c5bf88937c049dd66b6f868c9bbfd6f7",
  },
} as const;

const campaignPath = "tools/flash/src/campaign.rs";
const admissionPath = "tools/flash/src/campaign/admission.rs";
const markersPath = "tools/flash/src/campaign/markers.rs";
const soakPath = "tools/flash/src/campaign/markers/soak.rs";
const evidencePath = "tools/flash/src/campaign/evidence.rs";
const networkPath = "tools/flash/src/campaign/network/model.rs";
const lifecycleTestPath = "crates/bitaxe-stratum/src/v1/production_session/tests/lifecycle.rs";
const campaignTestPath = "tools/flash/src/tests/campaign.rs";
const criteriaPaths = [
  campaignPath,
  admissionPath,
  markersPath,
  soakPath,
  evidencePath,
  networkPath,
  lifecycleTestPath,
  campaignTestPath,
] as const;

const sourceFragments = new Map<string, readonly string[]>([
  [campaignPath, [
    "const MINING_DURATION_SECONDS: u64 = 600;",
    "let cleanup_result = environment.finish_usb_session();",
  ]],
  [admissionPath, [
    "command.board != BoardId::Ultra205 || !command.redact_evidence",
    "MiningCampaignStage::LiveShare | MiningCampaignStage::Soak => MINING_DURATION_SECONDS,",
    "command.profile == Some(MiningCampaignProfile::UpstreamDefault)",
  ]],
  [markersPath, [
    "marker.safe_stop == SafeStopMarker::Confirmed;",
    "assess_soak_terminal(marker, admission.duration_seconds)",
  ]],
  [soakPath, [
    "marker.accepted_share_count == 0",
    "marker.active_ms < duration_seconds.saturating_mul(1_000)",
  ]],
  [evidencePath, [
    "set_private_directory_mode(root)?;",
    "write_private_new_bytes(&paths.result, &result_bytes)",
    "redacted: true,",
  ]],
  [networkPath, [
    "self.close_elapsed_windows(600_000, serial);",
    "covered_window_count == REQUIRED_WINDOWS",
    "watchdog_valid: self.watchdog_valid,",
    "terminal_http_valid: self.terminal_http_valid,",
    "terminal_websocket_valid: self.terminal_websocket_valid,",
  ]],
  [lifecycleTestPath, ["fn active_duration_counts_from_authorized_mining()"]],
  [campaignTestPath, ["fn soak_requires_full_active_duration()"]],
]);

export class MiningCriteriaEvidenceError extends Error {
  public constructor(
    public readonly category: FailureCategory,
    message: string,
    public readonly publicValue: Readonly<Record<string, unknown>>,
  ) {
    super(message);
    this.name = "MiningCriteriaEvidenceError";
  }
}

function failure(category: FailureCategory, message: string): MiningCriteriaEvidenceError {
  return new MiningCriteriaEvidenceError(category, message, {
    stage: "sealed_mining_criteria_projection",
    hardware_rerun_used: false,
    terminal_attempt_reopened: false,
    projection_published: false,
  });
}

function sha256(value: string | Buffer): string {
  return createHash("sha256").update(value).digest("hex");
}

function lowerHex(value: string, length: number): boolean {
  return value.length === length && new RegExp(`^[0-9a-f]{${String(length)}}$`, "u").test(value);
}

async function childText(
  processPort: ProcessPort,
  program: string,
  args: readonly string[],
  context: string,
): Promise<string> {
  try {
    const outcome = await processPort.run(internalCommandSpec(program, [...args], (value) => value));
    if (outcome.timedOut || outcome.exitCode !== 0) {
      throw failure("evidence_invalid", `${context} did not pass`);
    }
    return outcome.stdout.trim();
  } catch (error) {
    if (error instanceof MiningCriteriaEvidenceError) throw error;
    throw failure("process_failed", `${context} launch failed`);
  }
}

async function requireAbsent(candidate: string, context: string): Promise<void> {
  try {
    await stat(candidate);
    throw failure("evidence_invalid", `${context} must be absent before projection`);
  } catch (error) {
    if (error instanceof MiningCriteriaEvidenceError) throw error;
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
}

function requireExactLines(document: string, lines: readonly string[], label: string): void {
  const actual = document.split(/\r?\n/u);
  for (const line of lines) {
    if (actual.filter((candidate) => candidate === line).length !== 1) {
      throw failure("evidence_invalid", `${label} facts are incomplete`);
    }
  }
}

function validateHistoricalFacts(summary: string, smoke: string, soak: string): void {
  requireExactLines(summary, [
    "phase21_status: passed",
    "phase21_evidence_closure: approved_controlled_no_share_soak",
    "redaction_status: passed",
    "raw_artifacts_committed: no",
    "reference_clean: passed",
  ], "Phase 21 summary");
  requireExactLines(smoke, [
    "live_mining_smoke_status: controlled-no-share",
    "controlled_package_boot_status: trusted",
    "controlled_runtime_harness_status: observed",
    "pool_lifecycle_status: active",
    "subscribe_status: sent",
    "authorize_status: sent",
    "notify_job_status: accepted work_enqueued=true",
    "bm1366_work_dispatch_status: typed_action_ready",
    "result_receive_status: bounded_no_result",
    "share_submission_status: bounded_no_share",
    "api_websocket_telemetry_update_status: ready",
    "watchdog_status: bounded observations present",
    "safe_stop_status: complete mining=disabled hardware_control=disabled work_submission=disabled",
    "redaction_status: passed",
  ], "Phase 21 smoke");
  requireExactLines(soak, [
    "bounded_soak_status: approved_controlled_no_share_soak",
    "duration_seconds: 300",
    "live_smoke_prerequisite: controlled-no-share",
    "controlled_package_boot_status: trusted",
    "controlled_runtime_harness_status: observed",
    "watchdog_responsiveness_status: passed",
    "api_snapshot_status: redacted_sample_captured",
    "websocket_frame_status: passed frames=5",
    "safe_stop_status: complete mining=disabled hardware_control=disabled work_submission=disabled",
    "redaction_status: passed",
  ], "Phase 21 soak");
}

function parseCoordinator(document: string): ProtocolCoordinatorEvidence {
  let value: unknown;
  try {
    value = JSON.parse(document);
  } catch {
    throw failure("evidence_invalid", "protocol coordinator projection is malformed");
  }
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw failure("evidence_invalid", "protocol coordinator projection must be an object");
  }
  return value as ProtocolCoordinatorEvidence;
}

function validateCoordinatorFacts(coordinator: ProtocolCoordinatorEvidence): void {
  if (coordinator.schema_version !== "bitaxe-protocol-coordinator-evidence-v1"
    || coordinator.board !== 205
    || !coordinator.coordinator.single_owner_serialization
    || !coordinator.coordinator.authorized_before_asic_dispatch
    || !coordinator.coordinator.qualified_result_before_submit
    || !coordinator.coordinator.accepted_submit_observed
    || !coordinator.coordinator.ordered_terminal_safe_stop
    || !coordinator.coordinator.watchdog_feed_in_owner_loop
    || !coordinator.coordinator.lifecycle_spans_compatible
    || coordinator.hardware_rerun_used
    || coordinator.redaction_status !== "passed") {
    throw failure("evidence_invalid", "protocol coordinator facts are incomplete");
  }
}

function requireUniqueFragments(document: string, fragments: readonly string[]): void {
  for (const fragment of fragments) {
    const first = document.indexOf(fragment);
    if (first === -1 || document.indexOf(fragment, first + fragment.length) !== -1) {
      throw failure("evidence_invalid", "mining criteria semantic fragment is not unique");
    }
  }
}

function requireOrderedFragments(document: string, fragments: readonly string[]): void {
  let cursor = 0;
  for (const fragment of fragments) {
    const index = document.indexOf(fragment, cursor);
    if (index === -1) throw failure("evidence_invalid", "mining criteria ordering is incomplete");
    cursor = index + fragment.length;
  }
}

async function validateCurrentSource(
  processPort: ProcessPort,
  gitProgram: string,
  currentSourceCommit: string,
): Promise<void> {
  const documents = await Promise.all(criteriaPaths.map(async (sourcePath) => [
    sourcePath,
    await childText(processPort, gitProgram, ["show", `${currentSourceCommit}:${sourcePath}`],
      "mining criteria source admission"),
  ] as const));
  const byPath = new Map<string, string>(documents);
  for (const [sourcePath, fragments] of sourceFragments) {
    const document = byPath.get(sourcePath);
    if (document === undefined) throw failure("evidence_invalid", "mining criteria source is missing");
    requireUniqueFragments(document, fragments);
  }
  requireOrderedFragments(byPath.get(admissionPath) ?? "", sourceFragments.get(admissionPath) ?? []);
  requireOrderedFragments(byPath.get(soakPath) ?? "", sourceFragments.get(soakPath) ?? []);
}

async function readAdmitted(
  workspaceRoot: string,
  sourcePath: string,
  expectedPath: string,
  expectedSha256: string,
  label: string,
): Promise<{ readonly resolved: string; readonly document: string; readonly sha256: string }> {
  const resolved = assertWithinWorkspace(workspaceRoot, sourcePath);
  if (path.relative(workspaceRoot, resolved) !== expectedPath) {
    throw failure("evidence_invalid", `${label} path is invalid`);
  }
  const document = await readFile(resolved, "utf8");
  const documentSha256 = sha256(document);
  if (documentSha256 !== expectedSha256) throw failure("evidence_invalid", `${label} digest is invalid`);
  return { resolved, document, sha256: documentSha256 };
}

export async function projectMiningCriteriaEvidence(
  workspaceRoot: string,
  options: MiningCriteriaEvidenceOptions,
  processPort: ProcessPort,
  gitProgram: string,
  validators: MiningCriteriaValidators,
  admittedDigests: MiningCriteriaAdmittedDigests = {
    summary: sourceSpecifications.summary.sha256,
    smoke: sourceSpecifications.smoke.sha256,
    soak: sourceSpecifications.soak.sha256,
    coordinator: sourceSpecifications.coordinator.sha256,
  },
): Promise<MiningCriteriaEvidence> {
  const projection = assertWithinWorkspace(workspaceRoot, options.projection);
  const candidate = `${projection}.candidate`;
  await requireAbsent(projection, "public projection");
  await requireAbsent(candidate, "projection candidate");

  const [summary, smoke, soak, coordinator] = await Promise.all([
    readAdmitted(workspaceRoot, options.summary, sourceSpecifications.summary.path,
      admittedDigests.summary, "Phase 21 summary"),
    readAdmitted(workspaceRoot, options.smoke, sourceSpecifications.smoke.path,
      admittedDigests.smoke, "Phase 21 smoke"),
    readAdmitted(workspaceRoot, options.soak, sourceSpecifications.soak.path,
      admittedDigests.soak, "Phase 21 soak"),
    readAdmitted(workspaceRoot, options.coordinatorProjection, sourceSpecifications.coordinator.path,
      admittedDigests.coordinator, "protocol coordinator projection"),
  ]);
  validateHistoricalFacts(summary.document, smoke.document, soak.document);
  await childText(processPort, validators.coordinator, [coordinator.resolved],
    "protocol coordinator source validation");
  validateCoordinatorFacts(parseCoordinator(coordinator.document));

  const [currentSourceCommit, referenceCommit] = await Promise.all([
    childText(processPort, gitProgram, ["rev-parse", "HEAD"], "current source identity"),
    childText(processPort, gitProgram,
      ["-C", path.join(workspaceRoot, "reference/esp-miner"), "rev-parse", "HEAD"],
      "reference source identity"),
  ]);
  if (!lowerHex(currentSourceCommit, 40) || !lowerHex(referenceCommit, 40)) {
    throw failure("evidence_invalid", "mining criteria source identity is invalid");
  }
  await validateCurrentSource(processPort, gitProgram, currentSourceCommit);
  const dirty = await childText(processPort, gitProgram,
    ["status", "--porcelain", "--", ...criteriaPaths,
      ...Object.values(sourceSpecifications).map((source) => source.path)],
    "mining criteria worktree state");
  if (dirty !== "") throw failure("evidence_invalid", "mining criteria paths have uncommitted drift");

  const requestSha256 = sha256(JSON.stringify({
    command: "project-mining-criteria-evidence",
    summary: sourceSpecifications.summary.path,
    smoke: sourceSpecifications.smoke.path,
    soak: sourceSpecifications.soak.path,
    coordinator_projection: sourceSpecifications.coordinator.path,
    current_source_commit: currentSourceCommit,
    projection: path.relative(workspaceRoot, projection),
  }));
  const evidence: MiningCriteriaEvidence = {
    schema_version: "bitaxe-mining-criteria-evidence-v1",
    board: 205,
    current_source_commit: currentSourceCommit,
    reference_commit: referenceCommit,
    workflow: {
      schema_version: "bitaxe-workflow-identity-v1",
      command: "project-mining-criteria-evidence",
      request_sha256: requestSha256,
    },
    source: {
      phase21_summary_sha256: summary.sha256,
      phase21_summary_valid: true,
      phase21_smoke_sha256: smoke.sha256,
      phase21_smoke_valid: true,
      phase21_soak_sha256: soak.sha256,
      phase21_soak_valid: true,
      protocol_coordinator_sha256: coordinator.sha256,
      protocol_coordinator_valid: true,
    },
    criteria: {
      historical_smoke_controlled_no_share: true,
      historical_soak_duration_seconds: 300,
      historical_watchdog_passed: true,
      historical_telemetry_observed: true,
      historical_safe_stop_confirmed: true,
      current_duration_seconds: 600,
      upstream_default_profile_required: true,
      active_duration_accounting: true,
      full_duration_required: true,
      accepted_share_required: true,
      network_correlation_required: true,
      safe_stop_required: true,
      private_evidence_required: true,
      redaction_required: true,
      source_spans_compatible: true,
      terminal_attempt_reopened: false,
    },
    hardware_rerun_used: false,
    redaction_status: "passed",
  };

  await mkdir(path.dirname(projection), { recursive: true });
  try {
    await writeFile(candidate, `${JSON.stringify(evidence, null, 2)}\n`,
      { encoding: "utf8", flag: "wx", mode: 0o600 });
    await chmod(candidate, 0o600);
    await childText(processPort, validators.evidence, [candidate], "independent evidence validation");
    await rename(candidate, projection);
    await chmod(projection, 0o644);
  } catch (error) {
    await unlink(candidate).catch(() => undefined);
    throw error;
  }
  return evidence;
}

export function projectMiningCriteriaEvidenceFromInvocation(
  workspaceRoot: string,
  invocation: ParsedInvocation,
  processPort: ProcessPort,
  resolveTool: (root: string, relative: string) => string,
): Promise<MiningCriteriaEvidence> {
  return projectMiningCriteriaEvidence(workspaceRoot, {
    summary: optionValue(invocation, "--summary"),
    smoke: optionValue(invocation, "--smoke"),
    soak: optionValue(invocation, "--soak"),
    coordinatorProjection: optionValue(invocation, "--coordinator-projection"),
    projection: optionValue(invocation, "--projection"),
  }, processPort, "git", {
    coordinator: resolveTool(workspaceRoot,
      "crates/bitaxe-automation-contracts/validate_protocol_coordinator_evidence"),
    evidence: resolveTool(workspaceRoot,
      "crates/bitaxe-automation-contracts/validate_mining_criteria_evidence"),
  });
}

File: crates/bitaxe-automation-contracts/src/mining_criteria_evidence.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{AutomationCommand, WorkflowIdentity, MINING_CRITERIA_EVIDENCE_SCHEMA};

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct MiningCriteriaSourceEvidence {
    pub phase21_summary_sha256: String,
    pub phase21_summary_valid: bool,
    pub phase21_smoke_sha256: String,
    pub phase21_smoke_valid: bool,
    pub phase21_soak_sha256: String,
    pub phase21_soak_valid: bool,
    pub protocol_coordinator_sha256: String,
    pub protocol_coordinator_valid: bool,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct MiningCriteriaObservationEvidence {
    pub historical_smoke_controlled_no_share: bool,
    pub historical_soak_duration_seconds: u64,
    pub historical_watchdog_passed: bool,
    pub historical_telemetry_observed: bool,
    pub historical_safe_stop_confirmed: bool,
    pub current_duration_seconds: u64,
    pub upstream_default_profile_required: bool,
    pub active_duration_accounting: bool,
    pub full_duration_required: bool,
    pub accepted_share_required: bool,
    pub network_correlation_required: bool,
    pub safe_stop_required: bool,
    pub private_evidence_required: bool,
    pub redaction_required: bool,
    pub source_spans_compatible: bool,
    pub terminal_attempt_reopened: bool,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct MiningCriteriaEvidence {
    pub schema_version: String,
    pub board: u16,
    pub current_source_commit: String,
    pub reference_commit: String,
    pub workflow: WorkflowIdentity,
    pub source: MiningCriteriaSourceEvidence,
    pub criteria: MiningCriteriaObservationEvidence,
    pub hardware_rerun_used: bool,
    pub redaction_status: String,
}

impl MiningCriteriaEvidence {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.schema_version != MINING_CRITERIA_EVIDENCE_SCHEMA || self.board != 205 {
            return Err("mining criteria evidence schema or board is invalid");
        }
        if self.workflow.schema_version != "bitaxe-workflow-identity-v1"
            || self.workflow.command != AutomationCommand::ProjectMiningCriteriaEvidence
        {
            return Err("mining criteria workflow identity is invalid");
        }
        for commit in [
            self.current_source_commit.as_str(),
            self.reference_commit.as_str(),
        ] {
            if !is_lower_hex(commit, 40) {
                return Err("mining criteria source identity is invalid");
            }
        }
        for digest in [
            self.workflow.request_sha256.as_str(),
            self.source.phase21_summary_sha256.as_str(),
            self.source.phase21_smoke_sha256.as_str(),
            self.source.phase21_soak_sha256.as_str(),
            self.source.protocol_coordinator_sha256.as_str(),
        ] {
            if !is_lower_hex(digest, 64) {
                return Err("mining criteria digest is invalid");
            }
        }
        if !self.source.phase21_summary_valid
            || !self.source.phase21_smoke_valid
            || !self.source.phase21_soak_valid
            || !self.source.protocol_coordinator_valid
        {
            return Err("mining criteria source evidence is invalid");
        }
        let criteria = &self.criteria;
        if !criteria.historical_smoke_controlled_no_share
            || criteria.historical_soak_duration_seconds != 300
            || !criteria.historical_watchdog_passed
            || !criteria.historical_telemetry_observed
            || !criteria.historical_safe_stop_confirmed
            || criteria.current_duration_seconds != 600
            || !criteria.upstream_default_profile_required
            || !criteria.active_duration_accounting
            || !criteria.full_duration_required
            || !criteria.accepted_share_required
            || !criteria.network_correlation_required
            || !criteria.safe_stop_required
            || !criteria.private_evidence_required
            || !criteria.redaction_required
            || !criteria.source_spans_compatible
            || criteria.terminal_attempt_reopened
        {
            return Err("mining criteria observation is incomplete");
        }
        if self.hardware_rerun_used || self.redaction_status != "passed" {
            return Err("mining criteria effect or redaction boundary is invalid");
        }
        Ok(())
    }
}

fn is_lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn evidence() -> MiningCriteriaEvidence {
        MiningCriteriaEvidence {
            schema_version: MINING_CRITERIA_EVIDENCE_SCHEMA.to_owned(),
            board: 205,
            current_source_commit: "a".repeat(40),
            reference_commit: "b".repeat(40),
            workflow: WorkflowIdentity {
                schema_version: "bitaxe-workflow-identity-v1".to_owned(),
                command: AutomationCommand::ProjectMiningCriteriaEvidence,
                request_sha256: "c".repeat(64),
            },
            source: MiningCriteriaSourceEvidence {
                phase21_summary_sha256: "d".repeat(64),
                phase21_summary_valid: true,
                phase21_smoke_sha256: "e".repeat(64),
                phase21_smoke_valid: true,
                phase21_soak_sha256: "f".repeat(64),
                phase21_soak_valid: true,
                protocol_coordinator_sha256: "1".repeat(64),
                protocol_coordinator_valid: true,
            },
            criteria: MiningCriteriaObservationEvidence {
                historical_smoke_controlled_no_share: true,
                historical_soak_duration_seconds: 300,
                historical_watchdog_passed: true,
                historical_telemetry_observed: true,
                historical_safe_stop_confirmed: true,
                current_duration_seconds: 600,
                upstream_default_profile_required: true,
                active_duration_accounting: true,
                full_duration_required: true,
                accepted_share_required: true,
                network_correlation_required: true,
                safe_stop_required: true,
                private_evidence_required: true,
                redaction_required: true,
                source_spans_compatible: true,
                terminal_attempt_reopened: false,
            },
            hardware_rerun_used: false,
            redaction_status: "passed".to_owned(),
        }
    }

    #[test]
    fn complete_closed_projection_is_accepted() {
        // Arrange
        let candidate = evidence();

        // Act
        let result = candidate.validate();

        // Assert
        assert_eq!(result, Ok(()));
    }

    #[test]
    fn incomplete_current_criteria_is_rejected() {
        // Arrange
        let mut candidate = evidence();
        candidate.criteria.full_duration_required = false;

        // Act
        let result = candidate.validate();

        // Assert
        assert_eq!(result, Err("mining criteria observation is incomplete"));
    }

    #[test]
    fn hardware_rerun_or_invalid_source_is_rejected() {
        // Arrange
        let mut rerun = evidence();
        rerun.hardware_rerun_used = true;
        let mut invalid_source = evidence();
        invalid_source.source.phase21_soak_valid = false;

        // Act / Assert
        assert_eq!(
            rerun.validate(),
            Err("mining criteria effect or redaction boundary is invalid")
        );
        assert_eq!(
            invalid_source.validate(),
            Err("mining criteria source evidence is invalid")
        );
    }
}
